Sources/OpenCombine/Publishers/Publishers.FlatMap.swift

```diff
@@ -0,0 +1,402 @@
+    private func lockedDeactivateAndReturnChildToCancel() -> Set<ChildSubscriber> {
+        downstream = nil
+        downstreamDemand = .none
+        let result = childSubscribers
+        childSubscribers.removeAll()
+        upstreamSubscription = nil
+        return result
+    }
+
+    private func cancelChildren(_ childrenToCancel: Set<ChildSubscriber>) {
+        childrenToCancel.forEach { $0.cancel() }
+    }
+
+    private final func evaluateActiveSubscriptionsExecuting(lockedWork: () -> Void) {
+        let shouldProcessQueue: Bool = lock.do {
+            lockedWork()
+            if childSubscribers.isEmpty && upstreamSubscription == nil {
+                sendFinishedAfterDrainingQueue = true
+                if !queueIsBeingProcessed {
+                    queueIsBeingProcessed = true
+                    return true
+                }
+            }
+            return false
+        }
+
+        if shouldProcessQueue {
+            processQueue()
+        }
+    }
+
+    private func child(_ child: ChildSubscriber,
+                       completedWith completion: Subscribers.Completion<Failure>) {
+        switch completion {
+        case .finished:
+            removeActiveSubscription(forChild: child)
+        case .failure:
+            downstream?.receive(completion: completion)
+            deactivate()
+        }
+    }
+
+    private func removeActiveSubscription(forChild child: ChildSubscriber) {
+        evaluateActiveSubscriptionsExecuting { childSubscribers.remove(child) }
+    }
+
+    private func child(_ child: ChildSubscriber,
+                       receivedValue value: Child.Output) -> Subscribers.Demand {
+        // Obtain the lock in order to determine what to do
+        let (valueWasBuffered, processTheQueue): (Bool, Bool) = lock.do {
+            // Consider pending values to send when assessing downstreamDemand
+            if downstreamDemand - valuesToSend.count == .none {
+                valuesToSend.append((value, child))
+                return (valueWasBuffered: true, processTheQueue: false)
+            } else {
+                valuesToSend.append((value, nil))
+                if queueIsBeingProcessed {
+                    return (valueWasBuffered: false, processTheQueue: false)
+                }
+                queueIsBeingProcessed = true
+                return (valueWasBuffered: false, processTheQueue: true)
+            }
+        }
+
+        let demandResult = valueWasBuffered || demandForChild() == .unlimited
+            ? Subscribers.Demand.none : Subscribers.Demand.max(1)
+
+        if processTheQueue {
+            processQueue()
+        }
+
+        return demandResult
+    }
+
+    private func demandForChild() -> Subscribers.Demand {
+        return self.downstreamDemand == .unlimited ? .unlimited : .max(1)
+    }
+
+    private enum QueueWorkStatus {
+        case noWork
+        case sendFinish
+        case sendValues(values: ArraySlice<WorkType>)
+    }
```

Comment:
Maybe move this enum into the `processQueue` function?

Reply:
Good idea, but that causes a compile error followed by an llvm crash. Lol.

error: type 'QueueWorkStatus' cannot be nested in generic function 'processQueue()'
        enum QueueWorkStatus {

Reply:
Oops! I think Apple folks would appreciate a bug report on [bugs.swift.org](https://bugs.swift.org).

Reply:
But yeah, my bad, I forgot that nested types are not allowed in generic contexts.